File: libs/shape/src/main/kotlin/mono/shape/command/GeneralShapeCommands.kt
/*
 * Copyright (c) 2023, tuanchauict
 */

package mono.shape.command

import mono.graphics.geo.Rect
import mono.shape.ShapeManager
import mono.shape.connector.ShapeConnectorUseCase.getPointInNewBound
import mono.shape.extra.ShapeExtra
import mono.shape.shape.AbstractShape
import mono.shape.shape.Group
import mono.shape.shape.Line

class ChangeBound(private val target: AbstractShape, private val newBound: Rect) : Command() {
    override fun getDirectAffectedParent(shapeManager: ShapeManager): Group? =
        shapeManager.getGroup(target.parentId)

    override fun execute(shapeManager: ShapeManager, parent: Group) {
        val currentVersion = target.versionCode
        target.setBound(newBound)
        if (currentVersion == target.versionCode) {
            return
        }

        val connectors = shapeManager.connectorManager.getConnectors(target)
        for (connector in connectors) {
            val anchorPointUpdate =
                Line.AnchorPointUpdate(connector.anchor, connector.getPointInNewBound(newBound))
            connector.line.moveAnchorPoint(anchorPointUpdate, isReduceRequired = false)
        }

        parent.update { true }
    }
}

class ChangeExtra(
    private val target: AbstractShape,
    private val newExtra: ShapeExtra
) : Command() {
    override fun getDirectAffectedParent(shapeManager: ShapeManager): Group? =
        shapeManager.getGroup(target.parentId)

    override fun execute(shapeManager: ShapeManager, parent: Group) {
        val currentVersion = target.versionCode
        target.setExtra(newExtra)
        if (currentVersion != target.versionCode) {
            parent.update { true }
        }
    }
}

File: monosketch-svelte/src/lib/mono/shape/command/general-shape-commands.ts
/*
 * Copyright (c) 2024, tuanchauict
 */

import type { Rect } from "$libs/graphics-geo/rect";
import type { ShapeExtra } from "$mono/shape/extra/shape-extra";
import { type Command, ShapeManager } from "$mono/shape/shape-manager";
import type { AbstractShape } from "$mono/shape/shape/abstract-shape";
import type { Group } from "$mono/shape/shape/group";

/**
 * A [Command] for changing the bound of a shape.
 */
export class ChangeBound implements Command {
    constructor(
        private readonly target: AbstractShape,
        private readonly newBound: Rect,
    ) {
    }

    getDirectAffectedParent(shapeManager: ShapeManager): Group | null {
        return shapeManager.getGroup(this.target.parentId);
    }

    execute(_shapeManager: ShapeManager, parent: Group) {
        const currentVersion = this.target.versionCode;
        this.target.setBound(this.newBound);
        if (currentVersion === this.target.versionCode) {
            return;
        }

        parent.update(() => true);
    }
}

/**
 * A [Command] for changing the extra of a shape.
 */
export class ChangeExtra implements Command {
    constructor(
        private readonly target: AbstractShape,
        private readonly newExtra: ShapeExtra,
    ) {
    }

    getDirectAffectedParent(shapeManager: ShapeManager): Group | null {
        return shapeManager.getGroup(this.target.parentId);
    }

    execute(_shapeManager: ShapeManager, parent: Group) {
        const currentVersion = this.target.versionCode;
        this.target.setExtra(this.newExtra);
        if (currentVersion !== this.target.versionCode) {
            parent.update(() => true);
        }
    }
}
